clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubPublisherClient.java

```diff
@@ -0,0 +1,127 @@
+            throw new ClientErrorException(HttpURLConnection.HTTP_CONFLICT, "Publisher client is null", e);
+        }
+    }
+
+    /**
+     * Creates a new instance of PubSubPublisherClient where a Pub/Sub Publisher is initialized. The Publisher is based
+     * on a created TopicName, which follows the format: projects/projectId/topics/topic.
+     *
+     * @param projectId The Google project id to use.
+     * @param topic The topic to create the publisher for.
+     * @return An instance of a PubSubPublisherClient.
+     * @throws ClientErrorException if the initialization of the Publisher failed.
+     */
+    public static PubSubPublisherClient createShared(final String projectId, final String topic)
+            throws ClientErrorException {
+        return new PubSubPublisherClient(projectId, topic);
+    }
+
+    /**
+     * Shuts the publisher down and frees resources.
+     */
+    @Override
+    public void close() {
+        if (Thread.currentThread().isInterrupted()) {
+            final boolean interrupted = Thread.interrupted();
+            if (interrupted) {
+                log.debug("Thread was interrupted, thus interrupted state has been cleared.");
+            }
+        }
+        final Context currentContext = Vertx.currentContext();
+        if (currentContext == null) {
+            throw new IllegalStateException("Client is not running on a Vert.x Context");
+        } else {
+            currentContext.executeBlocking(blockingHandler -> {
+                if (publisher != null) {
+                    publisher.shutdown();
+                    try {
+                        publisher.awaitTermination(1, TimeUnit.MINUTES);
+                        blockingHandler.complete();
+                    } catch (InterruptedException e) {
+                        log.debug("Resources are not freed properly, error", e);
+                        Thread.currentThread().interrupt();
+                        blockingHandler.fail(e);
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Publishes a message to Pub/Sub and transfer the returned ApiFuture into a Future.
+     *
+     * @param pubsubMessage The message to publish.
+     * @return The messageId wrapped in a Future.
+     */
+    public Future<String> publish(final PubsubMessage pubsubMessage) {
+        final Promise<String> result = Promise.promise();
+        final ApiFuture<String> messageIdFuture = publisher.publish(pubsubMessage);
+        ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<>() {
+
+            public void onSuccess(final String messageId) {
+                result.handle(Future.succeededFuture(messageId));
+            }
+
+            public void onFailure(final Throwable t) {
+                log.debug("Error publishing messages to Pub/Sub", t);
+                result.handle(Future.failedFuture(t));
+            }
+        }, MoreExecutors.directExecutor());
```

Comment:
does this mean, that the publishing will be run on the current thread?

Reply:
yes

Reply:
sorry for the confusion, that was a misunderstanding. The publishing call from the Pub/Sub client is not blocking the current event-loop thread since it is used asynchron within the callback function.